docs/src/userguide/numpy_tutorial.rst

```diff
@@ -356,43 +340,83 @@ get by declaring the memoryviews as contiguous:
 
-    In [19]: %timeit -n2 -r3 convolve_py.naive_convolve(f, g)
-    3.9 s ± 12.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
-    In [20]: %timeit -n2 -r3 convolve_cy.naive_convolve(f, g)
-    3.12 s ± 15.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
-    In [21]: %timeit -n2 -r3 convolve_typed.naive_convolve(f, g)
-    13.8 s ± 122 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
-    In [22]: %timeit -n2 -r3 convolve_memview.naive_convolve(f, g)
-    13.5 ms ± 455 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
-    In [23]: %timeit -n2 -r3 convolve_index.naive_convolve(f, g)
-    7.57 ms ± 151 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
-    In [23]: %timeit -n2 -r3 convolve_contiguous.naive_convolve(f, g)
+    In [23]: %timeit convolve_contiguous.naive_convolve(f, g)
     7.2 ms ± 40.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
 
+We're now 541 times faster than the interpreted Python version.
 
 Making the function cleaner
 ===========================
 
 Declaring types can make your code quite verbose. If you don't mind
 Cython inferring the C types of your variables, you can use
-the `infer_types=True` compiler directive. It will save you quite a bit
-of typing.
+the ``infer_types=True`` compiler directive at the top of the file.
+It will save you quite a bit of typing.
+
+Note that since type declarations must happen at the top indentation level,
+Cython won't infer the type of variable declared for the first time
+in other indentation levels. It would change too much the meaning of
+our code. This is why, we must still declare manually the type of the
+``value`` variable.
 
-# explain here why value must be typed
+And actually, manually giving the type of the ``value`` variable will
+be useful when using fused types.
 
 .. literalinclude:: ../../examples/userguide/convolve_infer_types.pyx
     :linenos:
 
-# explain here why it is faster.
+We now do a speed test:
+
+.. sourcecode:: ipython
+
+    In [24]: %timeit convolve_infer_types.naive_convolve(f, g)
+    5.33 ms ± 72.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
+
+We're now 731 times faster than the interpreted Python version.
+
+# Explain the black magic of why it's faster.
 
 More generic code
 ==================
 
-# Explain here templated
+All those speed gains are nice, but adding types constrains our code.
+At the moment, it would mean that our function only work with
+NumPy arrays with the ``np.intc`` type. Is it possible to make our
+code work for multiple NumPy data types?
+
+Yes, with the help of a new feature called fused types.
+You can learn more about it at :ref:`this section of the documentation
+<fusedtypes>`.
+It is similar to C++ 's templates. It generates mutiple function declarations
+at compile time, and then chooses the right one at run-time based on the
+types of the arguments provided. It is also possible to check with
+``if-else`` statements what is the value of the fused type.
```

Comment:
"By comparing types in if-conditions, it is also possible to execute entirely different code paths depending on the specific data type."